tracer/src/Datadog.Trace/SpanContextPropagator.cs

```diff
@@ -26,9 +27,11 @@ internal class SpanContextPropagator
         private readonly CultureInfo _invariantCulture = CultureInfo.InvariantCulture;
         private readonly IDatadogLogger _log = DatadogLogging.GetLoggerFor<SpanContextPropagator>();
         private readonly ConcurrentDictionary<Key, string?> _defaultTagMappingCache = new();
+        private readonly Func<IReadOnlyDictionary<string, string?>?, string, IEnumerable<string?>> _readOnlyDictionaryValueGetterDelegate;
 
         private SpanContextPropagator()
         {
+            _readOnlyDictionaryValueGetterDelegate = (carrier, name) => carrier != null && carrier.TryGetValue(name, out var value) ? new[] { value } : Enumerable.Empty<string?>();
```

Comment:
It bugs me that we are creating 3 arrays every time we call `Extract` now, even though we only ever use the first value it produces

I see two obvious ways around that:

- Create a "discriminated union" style `readonly struct OneOrMany`, and return that instead of `IEnumerable<string?>`
- Create overload for `Extract<T>`/`ParseUInt64()` etc, that take a `Func<TCarrier, string, string?>` instead of `Func<TCarrier, string, IEnumerable<string?>`?

What do you think?

Reply:
Good catch! I agree with this. What do you think of introducing `StringValues`? It's a bigger change than your suggestions, but I'm hoping we can start using it in many places going forwards and it's more optimized on .NET Core. #2270